core/libp2p/host.hpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_HOST_HPP
+#define KAGOME_HOST_HPP
+
+#include <functional>
+
+#include <gsl/span>
+#include <outcome/outcome.hpp>
+#include "libp2p/config.hpp"
+#include "libp2p/multi/multiaddress.hpp"
+#include "libp2p/peer/peer_id.hpp"
+#include "libp2p/peer/peer_repository.hpp"
+#include "libp2p/peer/protocol.hpp"
+#include "libp2p/stream/stream.hpp"
+
+#include <boost/asio.hpp>
+
+namespace libp2p {
+
+  class Host {
+   public:
+    using StreamHandler = void(std::unique_ptr<stream::Stream>);
+
+    /**
+     * @brief Get identifier of this Host
+     */
+    peer::PeerId getId() const noexcept;
+
+    /**
+     * @brief Get list of addresses this Host listens on
+     */
+    gsl::span<const multi::Multiaddress> getListenAddresses();
+
+    /**
+     * @brief Let Host handle given {@param proto} protocol
+     * @param proto protocol to be handled
+     * @param handler callback that is executed when some other Host creates
+     * stream to our host with {@param proto} protocol.
+     */
+    void setProtocolHandler(const peer::Protocol &proto,
+                            const std::function<StreamHandler> &handler);
+
+    void setProtocolHandlerMatch(
+        std::string_view prefix,
+        const std::function<bool(const peer::Protocol &)> &predicate,
+        const std::function<StreamHandler> &handler);
+
+    /**
+     * @brief Initiates connection to the peer {@param p}. If connection exists,
```

Comment:
Why do we need to check each time whether connection already exists? Could you pls clarify the use case?

Reply:
We will not check each time. This function is used, for example, if we know we are going to need a connection with that particular peer in the future, and we want to initiate it now to make the process of, for instance, stream creation, faster in the future. 
Creating a new connection is an expensive operation, and this can save us some time and resources, even taking into account a possibility, that this connection can be closed before we actually use it.